drv/cosmo-hf/src/hf.rs

```diff
@@ -503,6 +608,76 @@ impl idl::InOrderHostFlashImpl for ServerImpl {
         }
     }
 
+    // This does a sha256 on the entire range _except_ sector0
+    // which is treted as `0xff`. We don't hash sector0 because
+    // that's where we store our persistent information and this
+    // function is used to check against files.
+    fn hash_significant_bits(
+        &mut self,
+        _: &RecvMessage,
+        dev: HfDevSelect,
+    ) -> Result<(), RequestError<HfError>> {
+        self.drv.check_flash_mux_state()?;
+        if self.hash.task.init_sha256().is_err() {
+            return Err(HfError::HashError.into());
+        }
+
+        match self.hash.state {
+            HashState::Hashing { .. } => return Err(HfError::HashError.into()),
+            _ => (),
+        }
+
+        // If we already have a valid hash for the slot don't bother
+        // starting again
+        match dev {
+            HfDevSelect::Flash0 => match self.hash.cached_hash0 {
+                SlotHash::Hash { .. } => return Ok(()),
+                _ => {
+                    self.hash.cached_hash0 = SlotHash::HashInProgress;
+                }
+            },
+            HfDevSelect::Flash1 => match self.hash.cached_hash1 {
+                SlotHash::Hash { .. } => return Ok(()),
+                _ => {
+                    self.hash.cached_hash1 = SlotHash::HashInProgress;
+                }
+            },
+        }
+
+        // Treat sector 0 as all `0xff`
+        let mut buf = [0u8; PAGE_SIZE_BYTES];
+        buf.fill(0xff);
+        for _ in (0..SECTOR_SIZE_BYTES).step_by(buf.len()) {
+            self.hash
+                .task
+                .update(buf.len() as u32, &buf)
+                .map_err(|_| RequestError::Runtime(HfError::HashError))?;
+        }
+
+        self.hash.state = HashState::Hashing {
+            dev,
+            addr: drv_hf_api::SECTOR_SIZE_BYTES,
```

Comment:
I think there's a mixup here between dev-relative addressing and absolute addressing.

Here, we always set the hash range as `SECTOR_SIZE_BYTES..SLOT_SIZE_BYTES`, and include a `dev`.  These are dev-relative addresses, not absolute addresses.

However, in `step_hash`, we call `hash_range_update` with these same address.  `hash_range_update` naively translates them to `FlashAddr` (i.e. a global address) with `FlashAddr::new`.  This seems wrong, because addresses with `HfDevSelect::Flash1` will incorrectly read from the flash0 address range.

(This is the whole point of `FlashAddr`; perhaps `FlashAddr::new` should be called `FlashAddr::i_promise_this_is_a_global_address`)

We should instead be using `Self::flash_addr_for(addr, dev)` to go from dev-relative to absolute addresses.